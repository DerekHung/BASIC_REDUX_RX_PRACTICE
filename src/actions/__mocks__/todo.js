export const ADD_NEW_TODO = 'ADD_NEW_TODO';
export const OPEN_EDIT = 'OPEN_EDIT';
export const UPDATE_ITEM = 'UPDATE_ITEM';
export const COMPLETE_ITEM = 'COMPLETE_ITEM';
export const REMOVE_ITEM = 'REMOVE_ITEM';
export const CHANGE_FILTER = 'CHANGE_FILTER';
export const addNewTodo = jest.fn().mockImplementation(res => {
    return {
        type: ADD_NEW_TODO,
        value: res
    }
})

export const completeItem = jest.fn().mockImplementation(key => {
    return {
        type: COMPLETE_ITEM,
        key
    }
})

export const removeItem = jest.fn().mockImplementation(key => {
    return {
        type: REMOVE_ITEM,
        key
    }
})

////  有時間再補完 componet => action 這段測試

export const openEdit = (key) => {
    return {
        type: OPEN_EDIT,
        key
    }
}

export const updateItem = (value, key) => {
    return {
        type: UPDATE_ITEM,
        value,
        key
    }
}



export const changeFilter = (filter) => {
    return {
        type: CHANGE_FILTER,
        filter
    }
}