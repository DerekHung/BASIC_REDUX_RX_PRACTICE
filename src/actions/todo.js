export const ADD_NEW_TODO = 'ADD_NEW_TODO';
export const OPEN_EDIT = 'OPEN_EDIT';
export const UPDATE_ITEM = 'UPDATE_ITEM';
export const COMPLETE_ITEM = 'COMPLETE_ITEM';
export const REMOVE_ITEM = 'REMOVE_ITEM';
export const CHANGE_FILTER = 'CHANGE_FILTER';

export const openEdit = (key) => {
    return {
        type: OPEN_EDIT,
        key
    }
}

export const addNewTodo = (value) => {
    return {
        type: ADD_NEW_TODO,
        value
    }
}

export const updateItem = (value, key) => {
    return {
        type: UPDATE_ITEM,
        value,
        key
    }
}

export const completeItem = (key) => {
    return {
        type: COMPLETE_ITEM,
        key
    }
}

export const removeItem = (key) => {
    return {
        type: REMOVE_ITEM,
        key
    }
}

export const changeFilter = (filter) => {
    return {
        type: CHANGE_FILTER,
        filter
    }
}