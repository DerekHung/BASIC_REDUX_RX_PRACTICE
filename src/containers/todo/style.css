.container {
    min-width: 320px;
    width: 100%;
    padding: 10px;
    overflow: auto;
}
.main-input {
    width: 100%;
    height: 3em;
    padding: 0 10px;
}
.inner-input {
    width: calc( 100% - 110px);
    border: 0;
    outline: 0;
    background: #f4f4f4;
    margin-left: 40px;
}
.filter-box {
    display: flex;
}
.filter-box span {
    flex: 1;
    border: 5px solid #eee;
    cursor: pointer;
    text-align: center;
}
.list li {
    width: 100%;
    height: 50px;
    display: flex;
    justify-content: space-between;
    font-size: 25px;
    text-align: center;
    padding: 5px;
    border: 1px solid #eee;
    background: #fff;
}
.content {
    flex: 1;
    overflow: hidden;
}
.list del {
    margin: 0 auto;
}
