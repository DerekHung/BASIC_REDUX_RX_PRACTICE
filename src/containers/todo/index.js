import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import Input from '../../components/input';
import { 
    openEdit,
    addNewTodo,
    updateItem,
    completeItem,
    removeItem,
    changeFilter
} from '../../actions/todo';
import { selectByFilter } from '../../reducers/selectors';
import './style.css';

class TodoWrapper extends PureComponent {
    _handleKeyPress(cb, ...args) {
        const e = args[args.length - 1];
        if( e.keyCode === 13 ){
            cb(e.target.value, ...args);
            e.target.value = '';
        } 
    }
    _renderElement(list){
        const { openEdit, updateItem, completeItem, removeItem } = this.props;
        return list.map((data, index) => {
            if( data.completed ) return <li key={data.key}><del>{data.value}</del></li>
            if( data.onEdit ) return <li  key={data.key}><input className="inner-input" defaultValue={data.value} onKeyUp={this._handleKeyPress.bind(this, updateItem, data.key)}/></li>
            return (
                <li key={data.key}>
                    <i className="icono-check" onClick={completeItem.bind(this, data.key)} />
                    <span className="content">{data.value}</span>
                    <span className="lazy">
                        <i className="icono-dropper" onClick={openEdit.bind(this, data.key)} />
                        <i className="icono-forbidden" onClick={removeItem.bind(this, data.key)} />
                    </span>
                </li>
            );
        })
    }
    render(){
        const { todoList, addNewTodo, changeFilter } = this.props;
        return(
            <div className="container">
                <Input
                    type="text"
                    placeHolder="請輸入文字..."
                    className="main-input"
                    onKeyUp={this._handleKeyPress.bind(this, addNewTodo)}
                />
                <div className="filter-box">
                    <span onClick={changeFilter.bind(this, 'ALL')}>ALL</span>
                    <span onClick={changeFilter.bind(this, 'COMPLETED')}>完成</span>
                    <span onClick={changeFilter.bind(this, 'ACTIVE')}>未完成</span>
                </div>
                <ul className="list">{ this._renderElement(todoList) }</ul>
            </div>
        );
    }
}

const mapStateToPorps = (state) => {
    return {
        todoList: selectByFilter(state)
    }
}

export default connect(mapStateToPorps,{openEdit,addNewTodo,updateItem,completeItem,removeItem,changeFilter})(TodoWrapper);