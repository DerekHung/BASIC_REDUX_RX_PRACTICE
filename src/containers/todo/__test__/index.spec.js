import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount, render, configure } from 'enzyme';
import TodoWrapper from '../index';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
// import store from '../../../store';
import { addNewTodo, completeItem, removeItem } from '../../../actions/todo';
import configureStore from 'redux-mock-store'


configure({ adapter: new Adapter() });

jest.mock('../../../actions/todo');

describe('render todo container', () => {
    const middlewares = []
    const mockStore = configureStore(middlewares)
    let store = {};
    
    beforeEach(() => {
        store = mockStore({ 
            todoReducer: {todolist: [
                {
                    key: 0,
                    value: 'testdefault',
                    onEdit: false,
                    completed: false
                },
                {
                    key: 1,
                    value: 'testEdit',
                    onEdit: true,
                    completed: false
                },
                {
                    key: 2,
                    value: 'testcomplete',
                    onEdit: false,
                    completed: true
                },
            ],
            count: 0,
            filter: 'ALL'}});
    });
    

    it('should match the snapshot', () => {
        const tree = renderer.create(<Provider store={store}><TodoWrapper /></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    
    it('should typeing text & when type enter should call the actionï¼Œclear the input value', () => {
        const wrapper = shallow(<TodoWrapper store={store} />).dive();
        const inputWrapper = wrapper.find('.main-input');
        inputWrapper.simulate('keyUp', {target:{value: '123'}, keyCode: 13});
        expect(addNewTodo).toHaveBeenCalledWith("123", {"keyCode": 13, "target": {"value": ""}}); 
    });

    it('should send the complete action when you click complete icon', () => {
        const wrapper = shallow(<TodoWrapper store={store} />).dive();
        const component = wrapper.find('.icono-check');
        component.simulate('click');
        expect(completeItem).toHaveBeenCalledWith(0);
    });

    it('should send the delete action when you click delete icon', () => {
        const wrapper = shallow(<TodoWrapper store={store} />).dive();
        const component = wrapper.find('.icono-forbidden');
        component.simulate('click');
        expect(removeItem).toHaveBeenCalledWith(0);
    });
    
});
