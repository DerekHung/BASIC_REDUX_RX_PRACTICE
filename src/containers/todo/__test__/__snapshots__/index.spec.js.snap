// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render todo container should match the snapshot 1`] = `
<div
  className="container"
>
  <input
    className="main-input"
    onChange={undefined}
    onKeyUp={[Function]}
    placeholder="請輸入文字..."
  />
  <div
    className="filter-box"
  >
    <span
      onClick={[Function]}
    >
      ALL
    </span>
    <span
      onClick={[Function]}
    >
      完成
    </span>
    <span
      onClick={[Function]}
    >
      未完成
    </span>
  </div>
  <ul
    className="list"
  >
    <li>
      <i
        className="icono-check"
        onClick={[Function]}
      />
      <span
        className="content"
      >
        testdefault
      </span>
      <span
        className="lazy"
      >
        <i
          className="icono-dropper"
          onClick={[Function]}
        />
        <i
          className="icono-forbidden"
          onClick={[Function]}
        />
      </span>
    </li>
    <li>
      <input
        className="inner-input"
        defaultValue="testEdit"
        onKeyUp={[Function]}
      />
    </li>
    <li>
      <del>
        testcomplete
      </del>
    </li>
  </ul>
</div>
`;
